File: frontend/src/stores/alerts.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/api/client'

export interface AlertRule {
    id: string
    user_id: string
    name: string
    alert_type: 'threshold' | 'spike' | 'no_data'
    metric: 'error_rate' | 'avg_latency' | 'request_count' | 'p95_latency'
    scope_type: 'all' | 'server' | 'gateway'
    scope_id?: string
    operator: 'above' | 'below'
    threshold: number
    duration_minutes: number
    notify_dashboard: boolean
    enabled: boolean
    created_at: string
    updated_at: string
}

export interface ActiveAlert {
    id: string
    rule_id: string
    rule_name: string
    alert_type: string
    metric: string
    scope_type: string
    scope_id?: string
    scope_name?: string
    triggered_at: string
    trigger_value: number
    threshold: number
    status: 'triggered' | 'acknowledged'
}

export interface AlertHistory {
    id: string
    rule_id: string
    triggered_at: string
    resolved_at?: string
    trigger_value: number
    status: 'triggered' | 'resolved' | 'acknowledged'
}

export interface CreateAlertRule {
    name: string
    alert_type: string
    metric: string
    scope_type: string
    scope_id?: string
    operator: string
    threshold: number
    duration_minutes: number
    notify_dashboard?: boolean
}

export interface UpdateAlertRule {
    name?: string
    alert_type?: string
    metric?: string
    scope_type?: string
    scope_id?: string
    operator?: string
    threshold?: number
    duration_minutes?: number
    notify_dashboard?: boolean
    enabled?: boolean
}

export const useAlertsStore = defineStore('alerts', () => {
    const rules = ref<AlertRule[]>([])
    const activeAlerts = ref<ActiveAlert[]>([])
    const loading = ref(false)
    const error = ref<string | null>(null)

    async function fetchRules() {
        loading.value = true
        error.value = null
        try {
            const response = await api.get<AlertRule[]>('/alerts')
            rules.value = response.data
        } catch {
            error.value = 'Failed to fetch alert rules'
            rules.value = []
        } finally {
            loading.value = false
        }
    }

    async function fetchActiveAlerts() {
        try {
            const response = await api.get<ActiveAlert[]>('/alerts/active')
            activeAlerts.value = response.data
        } catch {
            activeAlerts.value = []
        }
    }

    async function createRule(input: CreateAlertRule): Promise<AlertRule | null> {
        loading.value = true
        error.value = null
        try {
            const response = await api.post<AlertRule>('/alerts', input)
            rules.value.unshift(response.data)
            return response.data
        } catch {
            error.value = 'Failed to create alert rule'
            return null
        } finally {
            loading.value = false
        }
    }

    async function updateRule(id: string, input: UpdateAlertRule): Promise<AlertRule | null> {
        loading.value = true
        error.value = null
        try {
            const response = await api.put<AlertRule>(`/alerts/${id}`, input)
            const index = rules.value.findIndex(r => r.id === id)
            if (index !== -1) {
                rules.value[index] = response.data
            }
            return response.data
        } catch {
            error.value = 'Failed to update alert rule'
            return null
        } finally {
            loading.value = false
        }
    }

    async function deleteRule(id: string): Promise<boolean> {
        loading.value = true
        error.value = null
        try {
            await api.delete(`/alerts/${id}`)
            rules.value = rules.value.filter(r => r.id !== id)
            return true
        } catch {
            error.value = 'Failed to delete alert rule'
            return false
        } finally {
            loading.value = false
        }
    }

    async function toggleRule(id: string, enabled: boolean): Promise<boolean> {
        const result = await updateRule(id, { enabled })
        return result !== null
    }

    async function acknowledgeAlert(id: string): Promise<boolean> {
        try {
            await api.post(`/alerts/${id}/acknowledge`)
            const alert = activeAlerts.value.find(a => a.id === id)
            if (alert) {
                alert.status = 'acknowledged'
            }
            return true
        } catch {
            return false
        }
    }

    return {
        rules,
        activeAlerts,
        loading,
        error,
        fetchRules,
        fetchActiveAlerts,
        createRule,
        updateRule,
        deleteRule,
        toggleRule,
        acknowledgeAlert,
    }
})

File: backend/src/routes/alerts.rs
//! Alerts API Routes
//!
//! CRUD endpoints for alert rules and history

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    Json,
};
use std::sync::Arc;
use uuid::Uuid;

use crate::AppState;
use crate::middleware::auth::AuthUser;
use crate::services::alerts::{
    self, AlertRule, ActiveAlert, CreateAlertRule, UpdateAlertRule,
    AlertHistoryQuery, AlertHistoryResponse,
};
use crate::messages::error;

/// GET /api/alerts - List org's alert rules
pub async fn list_rules(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
) -> Result<Json<Vec<AlertRule>>, (StatusCode, String)> {
    let org_id = auth.org_id;

    let rules = alerts::list_rules(&state.db.pool, org_id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    Ok(Json(rules))
}

/// POST /api/alerts - Create new alert rule
pub async fn create_rule(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
    Json(input): Json<CreateAlertRule>,
) -> Result<(StatusCode, Json<AlertRule>), (StatusCode, String)> {
    let org_id = auth.org_id;

    // Validate input
    if !["threshold", "spike", "no_data"].contains(&input.alert_type.as_str()) {
        return Err((StatusCode::BAD_REQUEST, "Invalid alert_type".to_string()));
    }
    if !["error_rate", "avg_latency", "request_count", "p95_latency"].contains(&input.metric.as_str()) {
        return Err((StatusCode::BAD_REQUEST, "Invalid metric".to_string()));
    }
    if !["all", "server", "gateway"].contains(&input.scope_type.as_str()) {
        return Err((StatusCode::BAD_REQUEST, "Invalid scope_type".to_string()));
    }
    if !["above", "below"].contains(&input.operator.as_str()) {
        return Err((StatusCode::BAD_REQUEST, "Invalid operator".to_string()));
    }

    let rule = alerts::create_rule(&state.db.pool, org_id, input)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    Ok((StatusCode::CREATED, Json(rule)))
}

/// GET /api/alerts/:id - Get alert rule by ID
pub async fn get_rule(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<Json<AlertRule>, (StatusCode, String)> {
    let org_id = auth.org_id;

    let rule = alerts::get_rule(&state.db.pool, org_id, id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    match rule {
        Some(r) => Ok(Json(r)),
        None => Err((StatusCode::NOT_FOUND, "Alert rule not found".to_string())),
    }
}

/// PUT /api/alerts/:id - Update alert rule
pub async fn update_rule(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
    Path(id): Path<Uuid>,
    Json(input): Json<UpdateAlertRule>,
) -> Result<Json<AlertRule>, (StatusCode, String)> {
    let org_id = auth.org_id;

    // Validate input if provided
    if let Some(ref alert_type) = input.alert_type {
        if !["threshold", "spike", "no_data"].contains(&alert_type.as_str()) {
            return Err((StatusCode::BAD_REQUEST, "Invalid alert_type".to_string()));
        }
    }
    if let Some(ref metric) = input.metric {
        if !["error_rate", "avg_latency", "request_count", "p95_latency"].contains(&metric.as_str()) {
            return Err((StatusCode::BAD_REQUEST, "Invalid metric".to_string()));
        }
    }

    let rule = alerts::update_rule(&state.db.pool, org_id, id, input)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    match rule {
        Some(r) => Ok(Json(r)),
        None => Err((StatusCode::NOT_FOUND, "Alert rule not found".to_string())),
    }
}

/// DELETE /api/alerts/:id - Delete alert rule
pub async fn delete_rule(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, (StatusCode, String)> {
    let org_id = auth.org_id;

    let deleted = alerts::delete_rule(&state.db.pool, org_id, id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    if deleted {
        Ok(StatusCode::NO_CONTENT)
    } else {
        Err((StatusCode::NOT_FOUND, "Alert rule not found".to_string()))
    }
}

/// GET /api/alerts/active - Get active (triggered) alerts
pub async fn get_active_alerts(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
) -> Result<Json<Vec<ActiveAlert>>, (StatusCode, String)> {
    let org_id = auth.org_id;

    let alerts = alerts::get_active_alerts(&state.db.pool, org_id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    Ok(Json(alerts))
}

/// GET /api/alerts/history - Get alert history with pagination
pub async fn get_history(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
    Query(query): Query<AlertHistoryQuery>,
) -> Result<Json<AlertHistoryResponse>, (StatusCode, String)> {
    let org_id = auth.org_id;

    let history = alerts::list_history(&state.db.pool, org_id, query)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    Ok(Json(history))
}

/// POST /api/alerts/:id/acknowledge - Acknowledge an active alert
pub async fn acknowledge_alert(
    State(state): State<Arc<AppState>>,
    auth: AuthUser,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, (StatusCode, String)> {
    let org_id = auth.org_id;

    let acknowledged = alerts::acknowledge_alert(&state.db.pool, org_id, id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("{}: {}", error::DATABASE_ERROR, e)))?;

    if acknowledged {
        Ok(StatusCode::OK)
    } else {
        Err((StatusCode::NOT_FOUND, "Alert not found or already acknowledged".to_string()))
    }
}
